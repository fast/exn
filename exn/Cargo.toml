# Copyright 2025 tison <wander4096@gmail.com>
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

[package]
name = "exn"
version = "0.1.4"

description = "A context-aware concrete Error type built on `std::error::Error`."

edition.workspace = true
homepage.workspace = true
license.workspace = true
readme.workspace = true
repository.workspace = true

[dependencies]
rustversion = { version = "1.0" }

[dev-dependencies]
insta = { version = "1.43.2" }
parse-display = { version = "0.10.0" }

[lints]
workspace = true
