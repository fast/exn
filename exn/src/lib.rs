// Copyright 2025 tison <wander4096@gmail.com>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! A context-aware concrete Error type built on `std::error::Error`
//!
//! # Examples
//!
//! ```no_run
//! use exn::Exn;
//! use exn::Result;
//! use exn::ResultExt;
//! use exn::bail;
//!
//! // Errors can be enum but notably don't need to chain source error.
//! #[derive(Debug)]
//! enum AppError {
//!     Fatal { consequences: &'static str },
//!     Trivial,
//! }
//!
//! impl std::fmt::Display for AppError {
//!     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
//!         match self {
//!             AppError::Fatal { consequences } => write!(f, "fatal error: {consequences}"),
//!             AppError::Trivial => write!(f, "trivial error"),
//!         }
//!     }
//! }
//!
//! impl std::error::Error for AppError {}
//!
//! // Errors can also be a struct.
//! #[derive(Debug)]
//! struct LogicError(String);
//!
//! impl std::fmt::Display for LogicError {
//!     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
//!         write!(f, "logic error: {}", self.0)
//!     }
//! }
//!
//! impl std::error::Error for LogicError {}
//!
//! fn do_logic() -> Result<(), LogicError> {
//!     bail!(LogicError("0 == 1".to_string()));
//! }
//!
//! fn main() -> Result<(), AppError> {
//!     do_logic().or_raise(|| AppError::Fatal {
//!         consequences: "math no longer works",
//!     })?;
//!
//!     Ok(())
//! }
//! ```
//!
//! The above program will print an error message like:
//!
//! ```text
//! fatal error: math no longer works, at exn/src/lib.rs:73:20
//! │
//! ╰─▶ logic error: 0 == 1, at exn/src/lib.rs:69:9
//! ```

#![feature(error_generic_member_access)]
#![deny(missing_docs)]

#[rustversion::not(nightly)]
compile_error!(
    "This crate requires a nightly compiler. Please use `rustup default nightly` or `cargo +nightly`."
);

mod impls;
mod macros;
mod option;
mod result;

pub use self::impls::Exn;
pub use self::impls::ExnTree;
pub use self::option::OptionExt;
pub use self::result::Result;
pub use self::result::ResultExt;

/// A trait bound of the error type of [`Exn`].
pub trait Error: std::error::Error + std::any::Any + Send + Sync + 'static {}

impl<T> Error for T where T: std::error::Error + std::any::Any + Send + Sync + 'static {}
