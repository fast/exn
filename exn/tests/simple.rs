// Copyright 2025 tison <wander4096@gmail.com>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use exn::Exn;
use exn::OptionExt;
use exn::ResultExt;

#[derive(Debug)]
struct SimpleError(&'static str);

impl std::fmt::Display for SimpleError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl std::error::Error for SimpleError {}

#[test]
fn test_error_straightforward() {
    let e1 = Exn::new(SimpleError("E1"));
    let e2 = e1.raise(SimpleError("E2"));
    let e3 = e2.raise(SimpleError("E3"));
    let e4 = e3.raise(SimpleError("E4"));
    let e5 = e4.raise(SimpleError("E5"));
    insta::assert_debug_snapshot!(e5);
}

#[test]
fn test_error_tree() {
    let e1 = Exn::new(SimpleError("E1"));
    let e3 = e1.raise(SimpleError("E3"));

    let e9 = Exn::new(SimpleError("E9"));
    let e10 = e9.raise(SimpleError("E10"));

    let e11 = Exn::new(SimpleError("E11"));
    let e12 = e11.raise(SimpleError("E12"));

    let e5 = Exn::from_iter([e3, e10, e12], SimpleError("E5"));

    let e2 = Exn::new(SimpleError("E2"));
    let e4 = e2.raise(SimpleError("E4"));

    let e7 = Exn::new(SimpleError("E7"));
    let e8 = e7.raise(SimpleError("E8"));

    let e6 = Exn::from_iter([e5, e4, e8], SimpleError("E6"));
    insta::assert_debug_snapshot!(e6);
}

#[test]
fn test_result_ext() {
    let result: Result<(), SimpleError> = Err(SimpleError("An error"));
    let result = result.or_raise(|| SimpleError("Another error"));
    insta::assert_debug_snapshot!(result.unwrap_err());
}

#[test]
fn test_option_ext() {
    let result: Option<()> = None;
    let result = result.ok_or_raise(|| SimpleError("An error"));
    insta::assert_debug_snapshot!(result.unwrap_err());
}

#[test]
fn test_from_error() {
    fn foo() -> exn::Result<(), SimpleError> {
        Err(SimpleError("An error"))?;
        Ok(())
    }

    let result = foo();
    insta::assert_debug_snapshot!(result.unwrap_err());
}

#[test]
fn test_bail() {
    fn foo() -> exn::Result<(), SimpleError> {
        exn::bail!(SimpleError("An error"));
    }

    let result = foo();
    insta::assert_debug_snapshot!(result.unwrap_err());
}

#[test]
fn test_ensure_ok() {
    fn foo() -> exn::Result<(), SimpleError> {
        exn::ensure!(true, SimpleError("An error"));
        Ok(())
    }

    foo().unwrap();
}

#[test]
fn test_ensure_fail() {
    fn foo() -> exn::Result<(), SimpleError> {
        exn::ensure!(false, SimpleError("An error"));
        Ok(())
    }

    let result = foo();
    insta::assert_debug_snapshot!(result.unwrap_err());
}
