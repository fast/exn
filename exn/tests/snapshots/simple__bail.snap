---
source: exn/tests/simple.rs
expression: result.unwrap_err()
---
An error, at exn/tests/simple.rs:91:9
